draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -448,12 +456,94 @@ Steps:
 20.  y = u * x (mod p)       // ux
 21.  y = y + v (mod p)       // ux + v
 22. Output (x, y)
-
 ~~~
 
 ### Shallue-Woestijne-Ulas Method {#swu}
 
-((TODO: write this section))
+The Shallue-Woestijne-Ulas (SWU) method, originated in part by
+Shallue and Woestijne {{SW06}} and later simplified and extended by Ulas {{SWU07}},
+deterministically encodes an artbirary string to a point on a curve. 
+This algorithm works for any curve over F_{p^n}. Given curve equation 
+g(x) = x^3 + Ax + B, two separate HashToBase implementations, H0 and H1, 
+this algorithm works as follows:
+
+~~~
+1. t = H0(alpha)
+2. u = H1(alpha)
+3. X1(t, u) = u
+4. X2(t, u) = (-B / A)(1 + 1 / (t^4 * g(u)^2 + t^2 * g(u)))
+5. X3(t, u) = t^3 * g(u)^2  * g(X2(t, u))
+6. If g(X1(t, u)) is square, output (X1(t, u), sqrt(g(X1(t, u))))
+7. If g(X2(t, u)) is square, output (X2(t, u), sqrt(g(X2(t, u))))
+8. Output (X3(t, u), sqrt(g(X3(t, u))))
+~~~
+
+The algorithm relies on the following equality:
+
+~~~
+t^3 * g(u)^2  * g(X2(t, u)) = g(X1(t, u)) * g(X2(t, u)) * g(X3(t, u))
+~~~
+
+Thus, at least one of X1(t, u), X2(t, u), X3(t, u) is the x-coordinate of a point
+on the curve. Computing the corresponding y value requires computing a square root.
+
+The following procedure implements this algorithm. It outputs a point with affine
+coordinates.
+
+~~~
+map2curve_squ(alpha)
+
+Input:
+
+  alpha - value to be hashed, an octet string
+  H0 - HashToBase implementation
+  H1 - HashToBase implementation
+
+Output:
+
+  (x, y) - a point in E
+
+Steps:
+
+1.    t = H0(alpha)   // {0,1}^* -> Fp
+2.    u = H1(alpha)   // {0,1}^* -> Fp
+3.   t2 = t^2
+4.   t4 = t2^2
+5.   gu = u^3
+6.   gu = gu + (A * u)
+7.   gu = gu + B      // gu = g(u)
+8.   x1 = u           // x1 = X1(t, u) = u
+9.   x2 = B * -1
+10.  x2 = x2 / A     
+11.  gx1 = x1^3
+12.  gx1 = gx1 + (A * x1)
+13.  gx1 = gx1 + B    // gx1 = g(X1(t, u))
+14.  d1 = gu^2
+15.  d1 = d1 * t4
+16.  d2 = t2 * gu
+17.  d3 = d1 + d2
+18.  d3 = d3^(-1)
+19.  n1 = 1 + d3
+20.  x2 = x2 * n1     // x2 = X2(t, u)
+21. gx2 = x2^3
+22. gx2 = gx2 + (A * x2)
+23. gx2 = gx2 + B     // gx2 = g(X2(t, u))
+24.  x3 = t2 * gu
+25.  x3 = x3 * x2     // x3 = X3(t, u)
+26. gx3 = x3^3
+27. gx3 = gx3 + (A * x3)
+28. gx3 = gx3 + B     // gx3 = g(X3(t, u))
+29.  l1 = gx1^((p - 1) / 2)
+30.  l2 = gx2^((p - 1) / 2)
+31.  s1 = gx1^(1/2)
+32.  s2 = gx2^(1/2)
+33.  s3 = gx3^(1/2)
+34. if l1 == 1:
```

Comment:
Should there be CMOVs here?

Reply:
Yeah, but I was going to do that for all algorithms in one swoop, in a different PR.